File: Data Structures & Algorithms/word-break/submission-3.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        seen = {}
        def helper(left, right):
            if (left, right) in seen:
                return seen[(left, right)]

            if right >= len(s):
                return left == len(s)
            
            if s[left : right + 1] in wordDict:
                seen[(left, right)] = True
                return helper(right + 1, right + 1) or helper(left, right + 1)
            seen[(left, right)] = False
            return helper(left, right + 1)
        
        return helper(0, 0)

File: Data Structures & Algorithms/spiral-matrix/submission-3.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();

        int top = 0;
        int bot = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;

        while (top <= bot && left <= right) {
            // top, left to right
            for (int col = left; col <= right; col++) {
                result.add(matrix[top][col]);
            }
            top++;

            // right, top to bot
            for (int row = top; row <= bot; row++) {
                result.add(matrix[row][right]);
            }
            right--;

            // bot, right to left
            if (top <= bot) {
                for (int col = right; col >= left; col--) {
                    result.add(matrix[bot][col]);
                }
                bot--;
            }
            
            // left, bot to top
            if (left <= right) {
                for (int row = bot; row >= top; row--) {
                    result.add(matrix[row][left]);
                }
                left++;
            }
        }
        return result;
    }
}
